north/src/runtime/linux/cgroups.rs

```diff
@@ -23,181 +23,93 @@ use std::{
-    },
-    #[error("Problem mounting cgroup: {context}")]
-    MountProblem {
-        context: String,
-        #[source]
-        error: Option<io::Error>,
-    },
-}
-
-#[async_trait::async_trait]
-trait CGroup: Sized {
-    /// Assign a PID to this cgroup
-    async fn assign(&self, pid: u32) -> Result<(), LinuxError> {
-        if !self.path().exists() {
-            Err(Error::InternalError(format!("Invalid cgroup {}", self.path().display())).into())
-        } else {
-            debug!("Assigning {} to {}", pid, self.path().display());
-            let tasks = self.path().join(TASKS);
-            write(&tasks, &pid.to_string()).await
-        }
-    }
-
-    /// Destroy the cgroup by removing the dir
-    async fn destroy(self) -> Result<(), Error> {
-        let path = self.path().to_owned();
-        if !path.exists() {
-            Err(Error::InternalError(format!(
-                "CGroup {} not found",
-                path.display()
-            )))
-        } else {
-            debug!("Destroying cgroup {}", path.display());
-            fs::remove_dir(&path).await.map_err(|e| Error::Destroy {
-                context: format!("Failed to remove {}", path.display()),
-                error: e,
-            })
-        }
-    }
-
-    /// Path to this cgroup instance
-    fn path(&self) -> &Path;
+    #[error("Failed to destroy cgroup {0}: {1}")]
+    Destroy(String, #[source] io::Error),
+    #[error("Cannot access mount points: {0}")]
+    MountPointsNotAccessible(#[source] io::Error),
+    #[error("No mount point for cgroup controller {0}")]
+    ControllerNotFound(String),
 }
 
 #[derive(Debug)]
-pub struct CGroupMem {
-    pub path: PathBuf,
-    monitor: mpsc::Sender<()>,
+struct CGroup {
+    cgroup: String,
+    path: PathBuf,
 }
 
-impl CGroup for CGroupMem {
-    fn path(&self) -> &Path {
-        self.path.as_path()
-    }
-}
-
-impl CGroupMem {
-    pub async fn new(
-        parent: &Path,
-        name: &str,
-        cgroup: &manifest::CGroupMem,
-        tx: EventTx,
-    ) -> Result<CGroupMem, LinuxError> {
-        let mount_point = get_mount_point("memory")?;
-        let path = mount_point.join(parent).join(name);
-        create(&path).await?;
-
-        // Configure memory limit
-        let limit_in_bytes = path.join(LIMIT_IN_BYTES);
-        debug!(
-            "Setting {} to {} bytes",
-            limit_in_bytes.display(),
-            cgroup.limit
-        );
-        write(&limit_in_bytes, &cgroup.limit.to_string()).await?;
-
-        // Configure oom
-        let oom_control = path.join(OOM_CONTROL);
-        write(&oom_control, "1").await?;
-
-        // Dropping monitor will stop the task below
-        let (monitor, mut rx) = mpsc::channel::<()>(1);
-        let name = name.to_string();
+pub async fn init_root_cgroups(config: &config::CGroups) -> Result<(), LinuxError> {
+    let mount_point = get_mount_point("cpuset")?;
+    let cpuset_root = mount_point.join(get_cgroup_root("cpuset", config));
```

Comment:
CGroups are now just a hashmap...so they are very generic. But you still treat it in a very non-generic way.
If they can indeed be generic, there should not be any code specific to a concrete cgroup anymore. If we need that anyway it should also be reflected in the datatype.

Reply:
I agree with you. The `cpuset` controller requires a "especial" treatment and that's why this function got introduced as a consequence. A possible solution to get rid of this would be to enable the possibility to configure the parent `cgroup` directories too, e.g. adding a `(param, value)` to the `cgroups` section in the `north.toml` configuration that are applied to the complete hierarchy. Does that sound reasonable?

Reply:
Sounds reasonable if we decide to expose the full cgroups configuration to the user.
I'm still not sure if that is a good thing though since it requires the user to fully understand cgroups to use north.
Might be a no-brainer for those pinguins but me for example I did not configure any cgroup before Northstar ;)
maybe we can discuss this here #234